### PR TITLE
Fix gdc

### PR DESCRIPTION
**Pull request by BarrOff**
Hello,

the configuraion of D-compilers gdc and ldc2 did not work for me. Therefore I made some changes, based on the files of dmd, fitting the parameters for each compiler. Now they pass the tests on my system.

If there is something wrong with those changes, please let me know and I will try to fix it.

### CODE REVIEW

File: xmake/modules/core/tools/gdc.lua [hunk 1]
```diff
@@ -11,15 +11,162 @@
...
+    local maps =
+    {
+        fast        = "-O"
+    ,   faster      = "-O -frelease"
+    ,   fastest     = "-O -frelease -fbounds-check=off"
+    ,   smallest    = "-O -frelease -fbounds-check=off"
+    ,   aggressive  = "-O -frelease -fbounds-check=off"
+    }
+    return maps[level]
+end
```
Comment (waruqi) [inline xmake/modules/core/tools/gdc.lua:50]: Most of the settings of ldc2/gdc are the same, you can inherit the configuration of dmd and only rewrite the configuration interface that needs to be modified. for example 

```lua
inherit("dmd")

function nf_optimize(self, level)
    local maps =
    {
        fast        = "-O"
    ,   faster      = "-O -frelease"
    ,   fastest     = "-O -frelease -fbounds-check=off"
    ,   smallest    = "-O -frelease -fbounds-check=off"
    ,   aggressive  = "-O -frelease -fbounds-check=off"
    }
    return maps[level]
end

function nf_includedir(self, dir)
    return "-I" .. os.args(dir)
end

-- make the link flag
function nf_link(self, lib)
    return "-l" .. lib
end
```

File: xmake/modules/core/tools/gdc.lua [hunk 2]
```diff
@@ -11,15 +11,162 @@
...
+    if self:has_flags("-rpath=" .. dir, "ldflags") then
+        return "-rpath=" .. os.args(dir:gsub("@[%w_]+", function (name)
+            local maps = {["@loader_path"] = "$ORIGIN", ["@executable_path"] = "$ORIGIN"}
+            return maps[name]
+        end))
+
+    elseif self:has_flags("-rpath " .. dir, "ldflags") then
+        return "-rpath " .. os.args(dir:gsub("%$ORIGIN", "@loader_path"))
+    end
+end
```
Comment (waruqi) [inline xmake/modules/core/tools/gdc.lua:132]: Do not repeat the same configuration, it is very redundant.

File: xmake/modules/core/tools/ldc2.lua [hunk 3]
```diff
@@ -11,28 +11,156 @@
...
 
     -- init shflags
-    self:set("dcshflags", "-shared")
+    self:set("dcshflags", "-shared", "--relocation-model=pic")
 
     -- init cxflags for the kind: shared
-    self:set("shared.dcflags", "")
+    self:set("shared.dcflags", "--relocation-model=pic")
+end
+
```
Comment (waruqi) [inline xmake/modules/core/tools/ldc2.lua:38]: Here too, we only need to rewrite init, not all.

```lua
inherit("dmd")

function init(self)
    _super.init(self)
    self:set("dcshflags", "-shared", "--relocation-model=pic")
    self:set("shared.dcflags", "--relocation-model=pic")
end
```

File: xmake/modules/detect/tools/gdc/has_flags.lua [hunk 4]
```diff
@@ -11,13 +11,113 @@
...
 -- Copyright (C) 2015-2020, TBOOX Open Source Group.
 --
--- @author      ruki
+-- @author      ruki, BarrOff
 -- @file        has_flags.lua
 --
 
 -- imports
-inherit("detect.tools.dmd.has_flags")
+import("lib.detect.cache")
```
Comment (waruqi) [inline xmake/modules/detect/tools/gdc/has_flags.lua:22]: I don't see the difference with `detect.tools.dmd.has_flags·, why not reuse it directly?

**Comment by BarrOff** [comment]
Hello,

this is the first time I am using Lua, however I have experience with multiple other languages.
Therefore I don't know about the inheritance mechanics of Lua. I tried using inheriting from dmd but it does not work completely.

In the gdc.lua file, if I remove any of the nf_syslink, link or compile functions, gdc is broken. However I did not make any changes to these functions, compared to the dmd versions.

Could you please help me with this?

**Comment by waruqi** [comment]
@BarrOff Abort gdc, we only need inherit `gcc` module. for example:

```lua

-- imports
inherit("gcc")

-- init it
function init(self)
    
    -- init arflags
    self:set("dcarflags", "-lib")

    -- init shflags
    self:set("dcshflags", "-shared", "-fPIC")

    -- init dcflags for the kind: shared
    self:set("shared.dcflags", "-fPIC")
end

-- make the optimize flag
function nf_optimize(self, level)
    local maps =
    {
        fast        = "-O"
    ,   faster      = "-O -frelease"
    ,   fastest     = "-O -frelease -fbounds-check=off"
    ,   smallest    = "-O -frelease -fbounds-check=off"
    ,   aggressive  = "-O -frelease -fbounds-check=off"
    }
    return maps[level]
end
```

It works for me.

**Comment by waruqi** [comment]
And for gdc2, we can only need:

```lua

-- imports
inherit("dmd")

-- init it
function init(self)

    -- init arflags
    self:set("dcarflags", "-lib")

    -- init shflags
    self:set("dcshflags", "-shared", "--relocation-model=pic")

    -- init cxflags for the kind: shared
    self:set("shared.dcflags", "--relocation-model=pic")
end

-- make the optimize flag
function nf_optimize(self, level)
    local maps =
    {
        fast        = "-O"
    ,   faster      = "-O --release"
    ,   fastest     = "-O --release --boundscheck=off"
    ,   smallest    = "-O --release --boundscheck=off"
    ,   aggressive  = "-O --release --boundscheck=off"
    }
    return maps[level]
end
```

**Comment by waruqi** [comment]
> In the gdc.lua file, if I remove any of the nf_syslink, link or compile functions, gdc is broken. 

This is because the compiling and linking commands and parameters of gdc are similar to gcc, but there are many differences with dmd, we should inherit the interface configuration from gcc.

**Comment by waruqi** [comment]
@BarrOff I merged some your patch codes and submit to the `dlang` branch. You can see https://github.com/xmake-io/xmake/pull/910

**Comment by waruqi** [comment]
I have merged https://github.com/xmake-io/xmake/pull/910 patch to dev branch. You can try it again. @BarrOff

**Pull request closed without merging**